src/core/components/pin.js

```diff
@@ -0,0 +1,431 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const CID = require('cids')
+const pinSet = require('./pin-set')
+const normalizeHashes = require('../utils').normalizeHashes
+const promisify = require('promisify-es6')
+const multihashes = require('multihashes')
+const Key = require('interface-datastore').Key
+const each = require('async/each')
+const waterfall = require('async/waterfall')
+const until = require('async/until')
+const once = require('once')
+
+const toB58String = multihashes.toB58String
+
+module.exports = function pin (self) {
+  let directPins = new Set()
+  let recursivePins = new Set()
+  let internalPins = new Set()
+
+  const pinDataStoreKey = new Key('/local/pins')
+
+  const repo = self._repo
+  const dag = self.dag
+
+  const pin = {
+    types: {
+      direct: 'direct',
+      recursive: 'recursive',
+      indirect: 'indirect',
+      internal: 'internal',
+      all: 'all'
+    },
+
+    clear: () => {
+      directPins.clear()
+      recursivePins.clear()
+      internalPins.clear()
+    },
+
+    set: pinSet(dag),
+
+    add: promisify((hashes, options, callback) => {
+      // callback (err, pinset)
+      if (typeof options === 'function') {
+        callback = options
+        options = null
+      }
+      callback = once(callback)
+      const recursive = !options || options.recursive !== false
+      normalizeHashes(self, hashes, (err, mhs) => {
+        if (err) { return callback(err) }
+        const result = {
+          // async result queue
+          payload: [],
+          update: (hash) => {
+            result.payload.push({hash})
+            if (result.payload.length === mhs.length) {
+              pin.flush((err, root) => {
+                if (err) { return callback(err) }
+                return callback(null, result.payload)
+              })
+            }
+          }
+        }
```

Comment:
This function seems a bit off. It is an async call that sets up `result`, which is going to be used below and when it gets called (result.update), it will be the call that calls the top-level callback.

If the section below needs that normalizeHashes to finish first, these steps should be wrapped into an `async.series` for better legibility and avoiding future errors (right now it is prone to race conditions)

Reply:
Agree it's convoluted, I'll work out a cleaner approach with async.